schema/manager.go

```diff
@@ -136,6 +136,68 @@ func (manager *Manager) OrderedSchemas() []*Schema {
 	return res
 }
 
+//reorderSchema tries reorder schemas using Tarjan's algorithm
+type state int
+
+const (
+	notVisited state = iota
+	visited
+	temporaryVisited
+)
+
+func visitSchema(schemas []*Schema, schemaList []string, schema *Schema, visitedSchema map[string]state) ([]string, error) {
+	if visitedSchema[schema.ID] == temporaryVisited {
+		return nil, fmt.Errorf("Schemas aren't DAG. We can't reorder automatically")
+	}
+	if visitedSchema[schema.ID] == visited {
+		return schemaList, nil
+	}
+	visitedSchema[schema.ID] = temporaryVisited
+	relatedSchemas := schema.relatedSchemas()
+	var err error
+	for _, relatedSchemaID := range relatedSchemas {
+		for _, relatedSchema := range schemas {
+			if relatedSchema.ID == relatedSchemaID {
+				schemaList, err = visitSchema(schemas, schemaList, relatedSchema, visitedSchema)
+				if err != nil {
+					return nil, err
+				}
+				break
+			}
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
+	visitedSchema[schema.ID] = visited
+	schemaList = append(schemaList, schema.ID)
+	return schemaList, nil
+}
+
+func reorderSchemas(schemas []*Schema) ([]string, error) {
+	var err error
+	schemaList := []string{}
+	visitedSchema := map[string]state{}
+	for _, schema := range schemas {
+		visitedSchema[schema.ID] = notVisited
+	}
+	for {
+		var schema *Schema
+		for _, s := range schemas {
+			if visitedSchema[s.ID] == notVisited {
+				schema = s
+			}
+		}
+		if schema == nil {
+			return schemaList, nil
+		}
+		schemaList, err = visitSchema(schemas, schemaList, schema, visitedSchema)
+		if err != nil {
+			return nil, err
```

Comment:
Let me make it clear that we don't care the situation that a subset of schemas is a DAG?